### PR TITLE
fix: broken Redis cache invalidation for non-long identifiers

### PR DESCRIPTION
**Pull request by netroms**

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java [hunk 1]
```diff
@@ -94,6 +118,30 @@ public void onPostDelete( PostDeleteEvent postDeleteEvent )
...
+        else if ( entity instanceof CompleteDataSetRegistration )
+        {
+            id = getCompleteDataSetRegistrationId( entity );
+        }
+        else if ( entity instanceof DataStatisticsEvent )
+        {
+            DataStatisticsEvent dataStatisticsEvent = (DataStatisticsEvent) entity;
+            id = dataStatisticsEvent.getId();
+        }
+        else if ( IdentifiableObject.class.isAssignableFrom( entity.getClass() ) )
```
Comment (jbee) [inline dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java:139]: can be `instanceof` check instead

File: dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/CacheInvalidationListener.java [hunk 2]
```diff
@@ -85,7 +97,7 @@ private void handleMessage( String message )
         if ( serverInstanceId.equals( uid ) )
         {
             log.debug( "Message came from this server, ignoring." );
-            return;
+            //            return;
```
Comment (jbee) [inline dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/CacheInvalidationListener.java:100]: Is this intentionally? If so remove the line entirely.

File: dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java [hunk 3]
```diff
@@ -114,6 +162,47 @@ private void publishMessage( Class<?> realClass, String message )
...
+    {
+        DataValue dataValue = (DataValue) entity;
+
+        long dataElementId = dataValue.getDataElement().getId();
+        long periodId = dataValue.getPeriod().getId();
+        long organisationUnitId = dataValue.getSource().getId();
+        long categoryOptionComboId = dataValue.getAttributeOptionCombo().getId();
+        long attributeOptionComboId = dataValue.getAttributeOptionCombo().getId();
+
+        return dataElementId + ";" + periodId + ";" + organisationUnitId + ";" + categoryOptionComboId + ";"
```
Comment (jbee) [inline dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java:175]: At some point we should think about using dedicated ID classes for this combination of IDs. Joining and splitting is relying on too much implicit knowledge. I guess we could equally well capture these in a small inner class like this? 
```java
@Value
static class DataValueId {
  long dataElementId;
  long periodId;
  long organisationUnitId;
  long categoryOptionComboId;
  long attributeOptionComboId;
}
```
Then you can simply do:
```java
return new DataValueId(dataElementId, periodId, organisationUnitId, categoryOptionComboId, attributeOptionComboId);
```
And so on for all the other IDs you need.

File: dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java [hunk 4]
```diff
@@ -114,6 +157,46 @@ private void publishMessage( Class<?> realClass, String message )
...
+    private Serializable getCompleteDataSetRegistrationId( Object entity )
+    {
+        CompleteDataSetRegistration completeDataSetRegistration = (CompleteDataSetRegistration) entity;
+
+        long dataSetId = completeDataSetRegistration.getDataSet().getId();
+        long periodId = completeDataSetRegistration.getPeriod().getId();
+        long organisationUnitId = completeDataSetRegistration.getSource().getId();
+        long categoryOptionComboId = completeDataSetRegistration.getAttributeOptionCombo().getId();
+
+        return dataSetId + ";" + periodId + ";" + organisationUnitId + ";" + categoryOptionComboId;
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java:197]: I think this part can be written better by joining e.g. with `String.join(";", List.of(dataSetId, periodId, ..))`.

File: dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java (hunk 3 above)
Comment (netroms) [reply dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java:170]: The problem here is that I need to serialize the IDs for the message passing through Redis. Without having to serialize and deserialize the messages, just using colon and semicolon separated IDs is by far the easiest.

File: dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java (hunk 4 above)
Comment (netroms) [reply dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java:193]: This does not work without having to cast to charseq

**Review by larshelge — APPROVED** [review]

**Pull request merged**